Source/NSConcretePointerFunctions.m

```diff
@@ -3,381 +3,297 @@
-      _x.hashFunction = hashString;
-      _x.isEqualFunction = equalString;
-    }
-  else if (options & NSPointerFunctionsStructPersonality)
-    {
-      _x.acquireFunction = acquireMallocMemory;
-      _x.descriptionFunction = describePointer;
-      _x.hashFunction = hashMemory;
-      _x.isEqualFunction = equalMemory;
+    int memoryOption = memoryType(options);
+    int personalityOption = personalityType(options);
+    BOOL copyIn = isCopyIn(options);
+
+    if (memoryOption == NSPointerFunctionsOpaqueMemory && copyIn) {
+        NSLog(@"ERROR: Unsupported use of NSPointerFunctionsCopyIn with NSPointerFunctionsOpaqueMemory");
+        NSParameterAssert(NO);
+        return nil;
     }
-  else if (options & NSPointerFunctionsIntegerPersonality)
-    {
-      _x.acquireFunction = acquireExistingMemory;
-      _x.descriptionFunction = describeInteger;
-      _x.hashFunction = hashDirect;
-      _x.isEqualFunction = equalDirect;
+
+    if (personalityOption == NSPointerFunctionsIntegerPersonality &&
+        memoryOption != NSPointerFunctionsOpaqueMemory) {
+        NSLog(@"ERROR: Unsupported use of NSPointerFunctionsIntegerPersonality with a memory type other than NSPointerFunctionsOpaqueMemory");
+        NSParameterAssert(NO);
+        return nil;
     }
-  else		/* objects */
-    {
-      if (options & NSPointerFunctionsZeroingWeakMemory)
-	{
-	  _x.acquireFunction = acquireExistingMemory;
-	}
-      else
-	{
-          _x.acquireFunction = acquireRetainedObject;
-	}
-      _x.descriptionFunction = describeObject;
-      _x.hashFunction = hashObject;
-      _x.isEqualFunction = equalObject;
+
+    switch (memoryOption) {
+        case NSPointerFunctionsStrongMemory: {
+            // Default option (0)
+            pf.acquireFunction = acquireRetainedObject;
+            pf.relinquishFunction = relinquishRetainedMemory;
+        } break;
+        case NSPointerFunctionsZeroingWeakMemory: {
+            NSLog(@"ERROR: Unsupported NSPointerFunctionsOptions option NSPointerFunctionsZeroingWeakMemory");
+            NSParameterAssert(NO);
+            return nil;
+        } break;
+        case NSPointerFunctionsOpaqueMemory: {
+            pf.acquireFunction = NULL;
+            pf.relinquishFunction = NULL;
+        } break;
+        case NSPointerFunctionsMallocMemory: {
+            NSLog(@"ERROR: Unsupported NSPointerFunctionsOptions option NSPointerFunctionsMallocMemory");
+            NSParameterAssert(NO);
+            return nil;
+        } break;
+        case NSPointerFunctionsMachVirtualMemory: {
+            NSLog(@"ERROR: Unsupported NSPointerFunctionsOptions option NSPointerFunctionsMachVirtualMemory");
+            NSParameterAssert(NO);
+            return nil;
+        } break;
+        case NSPointerFunctionsWeakMemory: {
+            pf.acquireFunction = NULL;
+            pf.relinquishFunction = NULL;
+        } break;
     }
 
+    switch (personalityOption) {
+        case NSPointerFunctionsObjectPersonality: {
```

Comment:
So to confirm, we do have incorrect usages of NSPointerArray in some places in libFantastical, because they do not specify the opaque personality when storing non-NSObject* types ?
Of course as we also talked, it may be the description/hash/equal functions never get invoked. But it's good to note regardless

Reply:
Yes, specifically in the ICS Parser it uses OpaqueMemory pointers (libical structs) but no personality is specified so it defaults to ObjectPersonality.